crypto/store/loader_file.c

```diff
@@ -0,0 +1,1027 @@
+ * Input:
+ *    pem_name:     If this blob comes from a PEM file, this holds
+ *                  the PEM name.  If it comes from another type of
+ *                  file, this is NULL.
+ *    pem_header:   If this blob comes from a PEM file, this holds
+ *                  the PEM headers.  If it comes from another type of
+ *                  file, this is NULL.
+ *    blob:         The blob of data to match with what this handler
+ *                  can use.
+ *    len:          The length of the blob.
+ *    handler_ctx:  For a handler marked repeatable, this pointer can
+ *                  be used to create a context for the handler.  IT IS
+ *                  THE HANDLER'S RESPONSIBILITY TO CREATE AND DESTROY
+ *                  THIS CONTEXT APPROPRIATELY, i.e. create on first call
+ *                  and destroy when about to return NULL.
+ *    ui_method:    Application UI method for getting a password, pin
+ *                  or any other interactive data.
+ *    ui_data:      Application data to be passed to ui_method when
+ *                  it's called.
+ * Output:
+ *    a STORE_INFO
+ */
+typedef STORE_INFO *(*STORE_FILE_try_decode_fn)(const char *pem_name,
+                                                const char *pem_header,
+                                                const unsigned char *blob,
+                                                size_t len, void **handler_ctx,
+                                                const UI_METHOD *ui_method,
+                                                void *ui_data);
+/*
+ * The eof function should return 1 if there's no more data to be found
+ * with the handler_ctx, otherwise 0.  This is only used when the handler is
+ * marked repeatable.
+ */
+typedef int (*STORE_FILE_eof_fn)(void *handler_ctx);
+/*
+ * The destroy_ctx function is used to destroy the handler_ctx that was
+ * intiated by a repeatable try_decode fuction.  This is only used when
+ * the handler is marked repeatable.
+ */
+typedef void (*STORE_FILE_destroy_ctx_fn)(void **handler_ctx);
+
+typedef struct store_file_handler_st {
+    const char *name;
+    STORE_FILE_try_decode_fn try_decode;
+    STORE_FILE_eof_fn eof;
+    STORE_FILE_destroy_ctx_fn destroy_ctx;
+
+    /* flags */
+    int repeatable;
+} STORE_FILE_HANDLER;
+
+static STORE_INFO *try_decode_PKCS12(const char *pem_name,
+                                     const char *pem_header,
+                                     const unsigned char *blob, size_t len,
+                                     void **pctx, const UI_METHOD *ui_method,
+                                     void *ui_data)
+{
+    STORE_INFO *store_info = NULL;
+    STACK_OF(STORE_INFO) *ctx = *pctx;
+
+    if (ctx == NULL) {
+        /* Initial parsing */
+        PKCS12 *p12;
+        int ok = 0;
+
+        if (pem_name != NULL)
+            /* No match, there is no PEM PKCS12 tag */
+            return NULL;
+
+        if ((p12 = d2i_PKCS12(NULL, &blob, len)) != NULL) {
```

Comment:
Not sure...  I *don't* check deliberately, for the possibility to have more than one blob in the same file.  Imagine for example a trusted cert file, with a X509 followed by a X509_AUX.  We currently only have that in PEM, but is there an actual reason not to have that possibility in raw DER?